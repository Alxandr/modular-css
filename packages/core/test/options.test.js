"use strict";

const path = require("path");

const dedent   = require("dedent");
const namer    = require("test-utils/namer.js");
const relative = require("test-utils/relative.js");
const { find } = require("test-utils/fixtures.js");

const Processor = require("../processor.js");

const sync = (css) => css.append({ selector : "a" });

const async = (css) => (
    new Promise((resolve) =>
        setTimeout(() => {
            sync(css);

            resolve();
        }, 0)
    )
);

describe("/processor.js", () => {
    describe("options", () => {
        describe("cwd", () => {
            it("should use an absolute path", async () => {
                const cwd = find("folder");
                const processor = new Processor({ cwd });
                
                const result = await processor.file("./folder.css");

                expect(processor.options.cwd).toBe(cwd);
                expect(result.file).toBe(find("folder.css"));
            });

            it("should accept a relative path but make it absolute", async () => {
                const cwd = path.relative(process.cwd(), find("folder"));
                const processor = new Processor({ cwd });
                
                const result = await processor.file("./folder.css");

                expect(processor.options.cwd).toBe(path.resolve(cwd));
                expect(result.file).toBe(find("folder.css"));
            });
        });

        describe("namer", () => {
            it("should use a custom naming function", async () => {
                const processor = new Processor({
                    namer : (filename, selector) =>
                        `${relative([ filename ])[0].replace(/[\/\.]/g, "_")}_${selector}`,
                });
                    
                const result = await processor.string(
                    "./packages/core/test/fixtures/simple.css",
                    ".wooga { }"
                );

                expect(result.exports).toMatchSnapshot();
                expect(result.details.text).toMatchSnapshot();
                expect(result.details.exports).toMatchSnapshot();
                expect(result.details.processed.root.toResult().css).toMatchSnapshot();
            });

            it("should require a namer if a string is passed", async () => {
                const processor = new Processor({
                    namer : "modular-css-namer",
                });
                    
                const result = await processor.string(
                    "./test/fixtures/simple.css",
                    ".wooga { }"
                );
                
                expect(result.exports).toMatchSnapshot();
            });

            it("should use the default naming function if a non-function is passed", async () => {
                const processor = new Processor({
                    namer : false,
                });
                    
                const result = await processor.string(
                    "./packages/core/test/fixtures/simple.css",
                    ".wooga { }"
                );
                
                expect(result.exports).toMatchSnapshot();
            });
        });

        describe("map", () => {
            it("should generate source maps", async () => {
                const processor = new Processor({
                    namer,
                    map : true,
                });
                
                await processor.file(find("start.css"));
                    
                const result = await processor.output({
                    from : "packages/core/test/fixtures/rewrite.css",
                    to   : "out.css",
                });

                expect(result.css).toMatchSnapshot();
            });

            it("should generate external source maps", async () => {
                const processor = new Processor({
                    namer,
                    map : {
                        internal : false,
                    },
                });
                
                await processor.file(find("start.css"));
                const result = await processor.output({
                    from : "packages/core/test/fixtures/rewrite.css",
                    to   : "out.css",
                });

                expect(result.css).toMatchSnapshot();
            });
        });

        describe("exportGlobals", () => {
            it("should not export :global values when exportGlobals is false", async () => {
                const processor = new Processor({
                    exportGlobals : false,
                });
                
                const result = await processor.string(
                    "./exportGlobals.css",
                    dedent(`
                        :global(.a) {}
                        .b {}
                    `)
                );

                expect(result.exports).toMatchSnapshot();
            });
        });

        describe("rewrite", () => {
            it("should rewrite url() references by default", async () => {
                const processor = new Processor();

                await processor.string(
                    "fixtures/rewrite.css",
                    dedent(`
                        .a {
                            background: url("img.png");
                        }
                    `)
                );

                const result = await processor.output({
                    from : "fixtures/rewrite.css",
                    to   : "output/rewrite.css",
                });
                
                expect(result.css).toMatchSnapshot();
            });

            it("should not rewrite url() references when falsey", async () => {
                const processor = new Processor({ rewrite : false });

                await processor.string(
                    "fixtures/rewrite.css",
                    dedent(`
                        .a {
                            background: url("img.png");
                        }
                    `)
                );

                const result = await processor.output({
                    from : "fixtures/rewrite.css",
                    to   : "./output/rewrite.css",
                });
                
                expect(result.css).toMatchSnapshot();
            });
            
            it("should pass through to postcss-url as config", async () => {
                const processor = new Processor({
                    rewrite : {
                        url : "inline",
                    },
                });
                
                await processor.file(find("rewrite-data-uri.css"));

                const result = await processor.output({
                    from : "fixtures/rewrite.css",
                    to   : "./output/rewrite.css",
                });
                
                expect(result.css).toMatchSnapshot();
            });
        });

        describe("lifecycle options", () => {
            describe("before", () => {
                it("should run sync postcss plugins before processing", async () => {
                    const processor = new Processor({
                        namer,
                        before : [ sync ],
                    });
                    
                    await processor.string(
                        "fixtures/sync-before.css",
                        ""
                    );

                    const result = await processor.output({
                        from : "fixtures/sync-before.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });

                it("should run async postcss plugins before processing", async () => {
                    const processor = new Processor({
                        namer,
                        before : [ async ],
                    });
                    
                    await processor.string(
                        "fixtures/async-before.css",
                        ""
                    );

                    const result = await processor.output({
                        from : "fixtures/async-before.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });
            });

            describe("processing", () => {
                it("should run sync postcss plugins processing processing", async () => {
                    const processor = new Processor({
                        namer,
                        processing : [ sync ],
                    });

                    await processor.string(
                        "fixtures/sync-processing.css",
                        ""
                    );
                    
                    const result = await processor.output({
                        from : "fixtures/sync-processing.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });

                it("should run async postcss plugins processing processing", async () => {
                    const processor = new Processor({
                        namer,
                        processing : [ async ],
                    });

                    await processor.string(
                        "fixtures/async-processing.css",
                        ""
                    );
                    
                    const result = await processor.output({
                        from : "fixtures/async-processing.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });

                it("should include exports from 'modular-css-export' modules", async () => {
                    const processor = new Processor({
                        namer,
                        processing : [ (css, result) => {
                            result.messages.push({
                                plugin  : "modular-css-exporter",
                                exports : {
                                    a : true,
                                    b : false,
                                },
                            });
                        } ],
                    });

                    const file = await processor.string(
                        "fixtures/async-processing.css",
                        ""
                    );

                    expect(file.exports).toMatchSnapshot();
                });
            });
            
            describe("after", () => {
                it("should use postcss-url by default", async () => {
                    const processor = new Processor();

                    await processor.file(find("relative.css"));
                    
                    const result = await processor.output({
                        from : "fixtures/rewrite.css",
                        to   : "./output/relative.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });
                
                it("should run sync postcss plugins", async () => {
                    const processor = new Processor({
                        namer,
                        after : [ sync ],
                    });

                    await processor.file(find("relative.css"));
                    
                    const result = await processor.output({
                        from : "fixtures/rewrite.css",
                        to   : "./output/relative.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });
                
                it("should run async postcss plugins", async () => {
                    const processor = new Processor({
                        namer,
                        after : [ async ],
                    });

                    await processor.file(find("relative.css"));
                    
                    const result = await processor.output({
                        from : "fixtures/rewrite.css",
                        to   : "./output/relative.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });
            });
            
            describe("done", () => {
                it("should run sync postcss plugins done processing", async () => {
                    const processor = new Processor({
                        namer,
                        done : [ sync ],
                    });
                    
                    await processor.string(
                        "fixtures/sync-done.css",
                        ""
                    );

                    const result = await processor.output({
                        from : "fixtures/sync-done.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });
                
                it("should run async postcss plugins done processing", async () => {
                    const processor = new Processor({
                        namer,
                        done : [ async ],
                    });
                    
                    await processor.string(
                        "fixtures/async-done.css",
                        ""
                    );

                    const result = await processor.output({
                        from : "fixtures/async-done.css",
                    });

                    expect(result.css).toMatchSnapshot();
                });
            });
        });
    });
});
