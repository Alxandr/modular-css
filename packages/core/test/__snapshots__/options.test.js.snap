// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/processor.js options exportGlobals should not export :global values when exportGlobals is false 1`] = `
Object {
  "b": Array [
    "mc8df92ddf_b",
  ],
}
`;

exports[`/processor.js options lifecycle options after should run async postcss plugins 1`] = `
"/* fixtures/relative.css */
.wooga {
    color: red;
    background: url(\\"../fixtures/folder/to.png\\");
}
a {}
"
`;

exports[`/processor.js options lifecycle options after should run sync postcss plugins 1`] = `
"/* fixtures/relative.css */
.wooga {
    color: red;
    background: url(\\"../fixtures/folder/to.png\\");
}
a {}
"
`;

exports[`/processor.js options lifecycle options after should use postcss-url by default 1`] = `
"/* fixtures/relative.css */
.mcf4554978_wooga {
    color: red;
    background: url(\\"../fixtures/folder/to.png\\");
}
"
`;

exports[`/processor.js options lifecycle options before should run async postcss plugins before processing 1`] = `
"/* fixtures/async-before.css */
a {}"
`;

exports[`/processor.js options lifecycle options before should run sync postcss plugins before processing 1`] = `
"/* fixtures/sync-before.css */
a {}"
`;

exports[`/processor.js options lifecycle options done should run async postcss plugins done processing 1`] = `
"/* fixtures/async-done.css */
a {}"
`;

exports[`/processor.js options lifecycle options done should run sync postcss plugins done processing 1`] = `
"/* fixtures/sync-done.css */
a {}"
`;

exports[`/processor.js options lifecycle options processing should include exports from 'modular-css-export' modules 1`] = `
Object {
  "a": true,
  "b": false,
}
`;

exports[`/processor.js options lifecycle options processing should run async postcss plugins processing processing 1`] = `
"/* fixtures/async-processing.css */
a {}"
`;

exports[`/processor.js options lifecycle options processing should run sync postcss plugins processing processing 1`] = `
"/* fixtures/sync-processing.css */
a {}"
`;

exports[`/processor.js options map should generate external source maps 1`] = `
"/* fixtures/folder/folder.css */
.folder { margin: 2px; }
/* fixtures/local.css */
.booga { background: green; }
/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZpeHR1cmVzL2ZvbGRlci9mb2xkZXIuY3NzIiwiZml4dHVyZXMvbG9jYWwuY3NzIiwiZml4dHVyZXMvc3RhcnQuY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLGdDQUFDO0FBQ0QsVUFBVSxZQUFZLEVBQUU7QUNEeEIsd0JBQUM7QUFJRCxTQUFTLGtCQUFrQixFQUFFO0FDSjdCLHdCQUFDO0FBRUQsU0RGQSxXQUFnQixDQUNoQixpQkFBaUIsRUNDc0I7QUFDdkMsU0ZIQSx3QkFBcUIsRUVHZSIsImZpbGUiOiJvdXQuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiQHZhbHVlIGZvbGRlcjogd2hpdGU7XG4uZm9sZGVyIHsgbWFyZ2luOiAycHg7IH1cbiIsIkB2YWx1ZSBvbmU6IHJlZDtcbkB2YWx1ZSB0d286IGJsdWU7XG5AdmFsdWUgZm9sZGVyIGZyb20gXCIuL2ZvbGRlci9mb2xkZXIuY3NzXCI7XG5cbi5ib29nYSB7IGJhY2tncm91bmQ6IGdyZWVuOyB9XG4ubG9vZ2EgeyBjb21wb3NlczogYm9vZ2E7IH1cbiIsIkB2YWx1ZSBvbmUsIHR3bywgZm9sZGVyIGZyb20gXCIuL2xvY2FsLmNzc1wiO1xuLndvb2dhIHsgY29tcG9zZXM6IGJvb2dhIGZyb20gXCIuL2xvY2FsLmNzc1wiOyB9XG4uYm9vZ2EgeyBjb2xvcjogb25lOyBiYWNrZ3JvdW5kOiB0d287IH1cbi50b29nYSB7IGJvcmRlcjogMXB4IHNvbGlkIGZvbGRlcjsgfVxuIl19 */"
`;

exports[`/processor.js options map should generate source maps 1`] = `
"/* fixtures/folder/folder.css */
.folder { margin: 2px; }
/* fixtures/local.css */
.booga { background: green; }
/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZpeHR1cmVzL2ZvbGRlci9mb2xkZXIuY3NzIiwiZml4dHVyZXMvbG9jYWwuY3NzIiwiZml4dHVyZXMvc3RhcnQuY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLGdDQUFDO0FBQ0QsVUFBVSxZQUFZLEVBQUU7QUNEeEIsd0JBQUM7QUFJRCxTQUFTLGtCQUFrQixFQUFFO0FDSjdCLHdCQUFDO0FBRUQsU0RGQSxXQUFnQixDQUNoQixpQkFBaUIsRUNDc0I7QUFDdkMsU0ZIQSx3QkFBcUIsRUVHZSIsImZpbGUiOiJvdXQuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiQHZhbHVlIGZvbGRlcjogd2hpdGU7XG4uZm9sZGVyIHsgbWFyZ2luOiAycHg7IH1cbiIsIkB2YWx1ZSBvbmU6IHJlZDtcbkB2YWx1ZSB0d286IGJsdWU7XG5AdmFsdWUgZm9sZGVyIGZyb20gXCIuL2ZvbGRlci9mb2xkZXIuY3NzXCI7XG5cbi5ib29nYSB7IGJhY2tncm91bmQ6IGdyZWVuOyB9XG4ubG9vZ2EgeyBjb21wb3NlczogYm9vZ2E7IH1cbiIsIkB2YWx1ZSBvbmUsIHR3bywgZm9sZGVyIGZyb20gXCIuL2xvY2FsLmNzc1wiO1xuLndvb2dhIHsgY29tcG9zZXM6IGJvb2dhIGZyb20gXCIuL2xvY2FsLmNzc1wiOyB9XG4uYm9vZ2EgeyBjb2xvcjogb25lOyBiYWNrZ3JvdW5kOiB0d287IH1cbi50b29nYSB7IGJvcmRlcjogMXB4IHNvbGlkIGZvbGRlcjsgfVxuIl19 */"
`;

exports[`/processor.js options namer should require a namer if a string is passed 1`] = `
Object {
  "wooga": Array [
    "AA",
  ],
}
`;

exports[`/processor.js options namer should use a custom naming function 1`] = `
Object {
  "wooga": Array [
    "packages_core_test_fixtures_simple_css_wooga",
  ],
}
`;

exports[`/processor.js options namer should use a custom naming function 2`] = `".wooga { }"`;

exports[`/processor.js options namer should use a custom naming function 3`] = `
Object {
  "wooga": Array [
    "packages_core_test_fixtures_simple_css_wooga",
  ],
}
`;

exports[`/processor.js options namer should use a custom naming function 4`] = `".packages_core_test_fixtures_simple_css_wooga { }"`;

exports[`/processor.js options namer should use the default naming function if a non-function is passed 1`] = `
Object {
  "wooga": Array [
    "mce31884e4_wooga",
  ],
}
`;

exports[`/processor.js options rewrite should not rewrite url() references when falsey 1`] = `
"/* fixtures/rewrite.css */
.mca5cd6948_a {
    background: url(\\"img.png\\");
}"
`;

exports[`/processor.js options rewrite should pass through to postcss-url as config 1`] = `
"/* fixtures/rewrite-data-uri.css */
.mcb65b22bc_a {
    background: url(\\"data:image/png;base64,iVBORwoaCgAAAA1JSERSAAAACgAAAAoIAgAAAAJQWOoAAAABc1JHQgCuzhzpAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwwAADsMBx2+oZAAAABl0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC4xOdTWsmQAAAAOSURBVChTYxgFJAMGBgABNgABY8OiGAAAAABJRU5ErkJggg==\\");
}
"
`;

exports[`/processor.js options rewrite should rewrite url() references by default 1`] = `
"/* fixtures/rewrite.css */
.mca5cd6948_a {
    background: url(\\"../fixtures/img.png\\");
}"
`;
